File: tests/test_trajectory.py
from __future__ import annotations

import math

import numpy as np
import pytest

import qupy as qp


def _z(qubit: int = 0) -> qp.Observable:
    return qp.observable_from_z(qp.Z(qubit))


def test_trajectory_expectation_matches_deterministic_noise() -> None:
    program = qp.x(qp.Program(1), 0)
    noisy = qp.NoisyProgram(
        program,
        [qp.NoiseInstruction(1, qp.bit_flip(0, 1.0))],
    )

    result = qp.trajectory_expectation(noisy, _z(), trajectories=32, seed=7)

    assert result.value == pytest.approx(1.0)
    assert result.standard_error == pytest.approx(0.0)
    assert result.trajectories == 32
    assert result.seed == 7
    assert result.state_bytes == 2 * np.dtype(np.complex128).itemsize
    assert result.exact is False
    assert result.backend == "native-cpu"
    assert result.method == "quantum-trajectory"


def test_trajectory_batch_matches_exact_density_with_reported_error() -> None:
    gamma = 0.25
    program = qp.x(qp.Program(1), 0)
    noisy = qp.NoisyProgram(
        program,
        [qp.NoiseInstruction(1, qp.amplitude_damping(0, gamma))],
    )

    result = qp.trajectory_expectations(
        noisy,
        [_z(), qp.observable([qp.pauli_term(2.0, [])])],
        trajectories=12000,
        seed=123456,
    )
    replay = qp.trajectory_expectations(
        noisy,
        [_z(), qp.observable([qp.pauli_term(2.0, [])])],
        trajectories=12000,
        seed=123456,
    )
    density = qp.density_matrix(noisy, backend="native-cpu")
    exact_z = float((density.values[0, 0] - density.values[1, 1]).real)

    assert result.values.shape == (2,)
    assert result.standard_errors.shape == (2,)
    assert result.values[0] == pytest.approx(exact_z, abs=0.035)
    assert result.values[1] == pytest.approx(2.0)
    assert 0.0 < result.standard_errors[0] < 0.025
    assert result.standard_errors[1] == pytest.approx(0.0)
    np.testing.assert_array_equal(result.values, replay.values)
    np.testing.assert_array_equal(result.standard_errors, replay.standard_errors)
    assert result.observable_count == 2


def test_trajectory_custom_kraus_and_insertion_order() -> None:
    reset_to_zero = qp.kraus_channel(
        0,
        [
            np.array([[1.0, 0.0], [0.0, 0.0]], dtype=np.complex128),
            np.array([[0.0, 1.0], [0.0, 0.0]], dtype=np.complex128),
        ],
    )
    program = qp.x(qp.Program(1), 0)
    reset = qp.NoisyProgram(program, [qp.NoiseInstruction(1, reset_to_zero)])
    before_gate = qp.NoisyProgram(
        program,
        [qp.NoiseInstruction(0, qp.bit_flip(0, 1.0))],
    )

    assert qp.trajectory_expectation(reset, _z(), 16, 5).value == pytest.approx(1.0)
    assert qp.trajectory_expectation(before_gate, _z(), 16, 5).value == pytest.approx(1.0)


def test_trajectory_state_memory_scales_as_statevector() -> None:
    qubits = 12
    noisy = qp.NoisyProgram(qp.Program(qubits), [])
    result = qp.trajectory_expectation(noisy, _z(), trajectories=2, seed=3)

    assert result.state_bytes == (1 << qubits) * np.dtype(np.complex128).itemsize
    density_bytes = (1 << (2 * qubits)) * np.dtype(np.complex128).itemsize
    assert result.state_bytes * (1 << qubits) == density_bytes


def test_trajectory_single_sample_and_validation() -> None:
    noisy = qp.NoisyProgram(qp.Program(1), [])
    single = qp.trajectory_expectation(noisy, _z(), trajectories=1, seed=9)

    assert single.value == pytest.approx(1.0)
    assert math.isnan(single.standard_error)

    with pytest.raises(ValueError, match="trajectory count"):
        qp.trajectory_expectation(noisy, _z(), trajectories=0, seed=1)
    with pytest.raises(ValueError, match="only native-cpu"):
        qp.trajectory_expectation(noisy, _z(), trajectories=8, seed=1, backend="native-cuda")
    with pytest.raises(ValueError, match="at least one observable"):
        qp.trajectory_expectations(noisy, [], trajectories=8, seed=1)

File: tests/cpp/test_trajectory.cpp
#include "qupy/trajectory.hpp"

#include <cmath>
#include <complex>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

void require(bool condition, const std::string& message) {
    if (!condition) {
        throw std::runtime_error(message);
    }
}

template <typename Function>
void require_invalid(Function&& function, const std::string& message) {
    try {
        function();
    } catch (const std::invalid_argument&) {
        return;
    }
    throw std::runtime_error(message);
}

qupy::Observable z_observable(std::size_t qubit) {
    return qupy::Observable({
        qupy::PauliTerm(1.0, {{qubit, qupy::Pauli::Z}}),
    });
}

void test_deterministic_unitary_and_noise() {
    qupy::Program program(1U);
    program = qupy::x(program, 0U);
    const qupy::NoisyProgram deterministic(
        program,
        {{1U, qupy::bit_flip(0U, 1.0)}}
    );
    const qupy::TrajectoryBatch result = qupy::trajectory_expectations(
        deterministic, {z_observable(0U)}, 32U, 7U
    );

    require(result.observable_count == 1U, "trajectory observable count mismatch");
    require(result.trajectories == 32U, "trajectory count mismatch");
    require(result.seed == 7U, "trajectory seed mismatch");
    require(result.state_bytes == 2U * sizeof(qupy::Complex), "trajectory state byte count mismatch");
    require(!result.exact, "trajectory estimate was marked exact");
    require(result.backend == "native-cpu", "trajectory backend mismatch");
    require(result.method == "quantum-trajectory", "trajectory method mismatch");
    require(std::abs(result.values.front() - 1.0) < 1e-12, "deterministic trajectory value mismatch");
    require(result.standard_errors.front() == 0.0, "deterministic trajectory error was nonzero");
}

void test_noise_insertion_order() {
    qupy::Program program(1U);
    program = qupy::x(program, 0U);
    const qupy::NoisyProgram before_gate(
        program,
        {{0U, qupy::bit_flip(0U, 1.0)}}
    );
    const auto result = qupy::trajectory_expectations(
        before_gate, {z_observable(0U)}, 8U, 11U
    );
    require(std::abs(result.values.front() - 1.0) < 1e-12, "noise insertion point ordering mismatch");
}

void test_amplitude_damping_converges_and_replays() {
    qupy::Program excited(1U);
    excited = qupy::x(excited, 0U);
    const double gamma = 0.25;
    const qupy::NoisyProgram noisy(
        excited,
        {{1U, qupy::amplitude_damping(0U, gamma)}}
    );
    const qupy::TrajectoryBatch first = qupy::trajectory_expectations(
        noisy, {z_observable(0U)}, 20000U, 123456U
    );
    const qupy::TrajectoryBatch replay = qupy::trajectory_expectations(
        noisy, {z_observable(0U)}, 20000U, 123456U
    );

    const double expected = 2.0 * gamma - 1.0;
    require(std::abs(first.values.front() - expected) < 0.03, "trajectory damping estimate did not converge");
    require(first.standard_errors.front() > 0.0, "stochastic trajectory error was not reported");
    require(first.standard_errors.front() < 0.02, "trajectory standard error is unexpectedly large");
    require(first.values == replay.values, "fixed-seed trajectory replay changed values");
    require(first.standard_errors == replay.standard_errors, "fixed-seed trajectory replay changed errors");
}

void test_custom_kraus_and_multiple_observables() {
    const qupy::NoiseChannel reset_to_zero = qupy::kraus_channel(
        0U,
        {
            {1.0, 0.0, 0.0, 0.0},
            {0.0, 1.0, 0.0, 0.0},
        }
    );
    qupy::Program excited(1U);
    excited = qupy::x(excited, 0U);
    const qupy::NoisyProgram noisy(excited, {{1U, reset_to_zero}});
    const qupy::Observable identity({qupy::PauliTerm(2.0, {})});
    const auto result = qupy::trajectory_expectations(
        noisy, {z_observable(0U), identity}, 16U, 5U, "cpu"
    );

    require(result.values.size() == 2U, "trajectory batch shape mismatch");
    require(std::abs(result.values[0] - 1.0) < 1e-12, "custom Kraus trajectory mismatch");
    require(std::abs(result.values[1] - 2.0) < 1e-12, "trajectory identity observable mismatch");
    require(result.standard_errors[0] == 0.0, "deterministic custom Kraus error was nonzero");
    require(result.standard_errors[1] == 0.0, "constant observable error was nonzero");
}

void test_single_trajectory_and_validation() {
    const qupy::NoisyProgram clean(qupy::Program(1U), {});
    const auto single = qupy::trajectory_expectations(
        clean, {z_observable(0U)}, 1U, 9U, "native-cpu"
    );
    require(std::abs(single.values.front() - 1.0) < 1e-12, "single trajectory value mismatch");
    require(std::isnan(single.standard_errors.front()), "single trajectory error should be undefined");

    require_invalid(
        [&] { static_cast<void>(qupy::trajectory_expectations(clean, {z_observable(0U)}, 0U, 1U)); },
        "zero trajectory count was accepted"
    );
    require_invalid(
        [&] { static_cast<void>(qupy::trajectory_expectations(clean, {}, 8U, 1U)); },
        "empty trajectory observable set was accepted"
    );
    require_invalid(
        [&] {
            static_cast<void>(qupy::trajectory_expectations(
                clean, {z_observable(0U)}, 8U, 1U, "native-cuda"
            ));
        },
        "unsupported trajectory backend was accepted"
    );
    require_invalid(
        [&] { static_cast<void>(qupy::trajectory_expectations(clean, {z_observable(1U)}, 8U, 1U)); },
        "out-of-range trajectory observable was accepted"
    );
}

}  // namespace

int main() {
    test_deterministic_unitary_and_noise();
    test_noise_insertion_order();
    test_amplitude_damping_converges_and_replays();
    test_custom_kraus_and_multiple_observables();
    test_single_trajectory_and_validation();
    return 0;
}
